File: backend/app/schemas/marine.py
"""Marine observation and condition schemas."""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field

class Coordinates(BaseModel):
    latitude: float = Field(..., description="Latitude in decimal degrees")
    longitude: float = Field(..., description="Longitude in decimal degrees")

class MarineObservation(BaseModel):
    location: Coordinates
    timestamp: str
    sst: Optional[float] = Field(None, description="Sea Surface Temperature in Celsius")
    chlorophyll: Optional[float] = Field(None, description="Chlorophyll-a concentration in mg/m3")
    wave_height: Optional[float] = Field(None, description="Significant wave height in meters")
    wave_direction: Optional[float] = Field(None, description="Wave direction in degrees")
    wind_speed: Optional[float] = Field(None, description="Wind speed in km/h")
    wind_direction: Optional[float] = Field(None, description="Wind direction in degrees")
    rainfall: Optional[float] = Field(None, description="Precipitation in mm/h")
    tide: Optional[str] = Field(None, description="Tide status (High/Low/Rising/Falling)")
    tide_height_m: Optional[float] = Field(None, description="Tide height in meters")
    sea_state: Optional[str] = Field(None, description="Douglas sea scale state description")
    source: str = "Demo Data Provider"
    data_type: str = "demo"  # 'observed' | 'forecast' | 'demo'
    is_demo: bool = True

class PFZZone(BaseModel):
    id: str
    name: str
    location: Coordinates
    polygon: Optional[List[List[float]]] = Field(default_factory=list, description="GeoJSON-style ring of [lat, lon]")
    distance_km: float = Field(..., description="Distance from querying vessel in km")
    bearing_deg: float = Field(..., description="Bearing in degrees from vessel")
    bearing_compass: str = Field(..., description="Compass direction (e.g. WNW)")
    sst_c: float
    chlorophyll_mg_m3: float
    suitability_score: float = Field(..., description="0 to 100 environmental index")
    safety_rating: str = Field(..., description="SAFE, CAUTION, AVOID")
    recommendation: str
    avoids: bool = False
    source: str = "INCOIS / ISRO Oceansat-3 (Synthetic Demo)"
    is_demo: bool = True

class WeatherReport(BaseModel):
    location: Coordinates
    timestamp: str
    temperature_c: float
    wind_speed_kmh: float
    wind_direction_deg: float
    wind_gust_kmh: float
    wave_height_m: float
    wave_direction_deg: float
    rainfall_mm: float
    humidity_pct: float
    visibility_km: float
    lightning_detected: bool
    lightning_distance_km: Optional[float] = None
    cyclone_status: str  # 'none' | 'watch' | 'warning'
    cyclone_category: Optional[str] = None
    advisory_text: Optional[str] = None
    source: str = "IMD / MOSDAC (Synthetic Demo)"
    is_demo: bool = True

File: frontend/src/types/marine.ts
export interface Coordinates {
  latitude: float;
  longitude: float;
}

export type float = number;

export interface MarineObservation {
  location: Coordinates;
  timestamp: string;
  sst?: number;
  chlorophyll?: number;
  wave_height?: number;
  wave_direction?: number;
  wind_speed?: number;
  wind_direction?: number;
  rainfall?: number;
  tide?: string;
  tide_height_m?: number;
  sea_state?: string;
  source: string;
  data_type: string;
  is_demo: boolean;
}

export interface WeatherReport {
  location: Coordinates;
  timestamp: string;
  temperature_c: number;
  wind_speed_kmh: number;
  wind_direction_deg: number;
  wind_gust_kmh: number;
  wave_height_m: number;
  wave_direction_deg: number;
  rainfall_mm: number;
  humidity_pct: number;
  visibility_km: number;
  lightning_detected: boolean;
  lightning_distance_km?: number;
  cyclone_status: 'none' | 'watch' | 'warning';
  cyclone_category?: string;
  advisory_text?: string;
  source: string;
  is_demo: boolean;
}

export interface PFZZone {
  id: string;
  name: string;
  location: Coordinates;
  polygon: [number, number][];
  distance_km: number;
  bearing_deg: number;
  bearing_compass: string;
  sst_c: number;
  chlorophyll_mg_m3: number;
  suitability_score: number;
  safety_rating: 'SAFE' | 'CAUTION' | 'AVOID';
  recommendation: string;
  avoids: boolean;
  source: string;
  is_demo: boolean;
}

export interface FactorScore {
  factor_name: string;
  raw_value: number;
  unit: string;
  score: number;
  weight: number;
  weighted_score: number;
  severity: 'LOW' | 'MODERATE' | 'HIGH' | 'EXTREME';
  explanation: string;
}

export interface RiskAssessment {
  overall_score: number;
  risk_level: 'LOW' | 'MODERATE' | 'HIGH' | 'EXTREME';
  safety_verdict: 'SAFE' | 'SAFE_WITH_CAUTION' | 'UNSAFE' | 'HAZARDOUS';
  recommendation: string;
  factors: FactorScore[];
  summary_reasons: string[];
  timestamp: string;
  calculation_method: string;
  disclaimer: string;
}

export interface Waypoint {
  name: string;
  latitude: number;
  longitude: number;
  hazard_distance_km?: number;
  segment_risk: string;
}

export interface RouteOption {
  route_type: 'shortest' | 'safe';
  waypoints: Waypoint[];
  distance_km: number;
  estimated_duration_hours: number;
  risk_level: string;
  hazards_intersected: string[];
  description: string;
}

export interface RouteComparison {
  origin: Coordinates;
  destination: Coordinates;
  shortest_route: RouteOption;
  safe_route: RouteOption;
  recommendation: string;
  reasoning: string;
}

export interface MarineAlert {
  id: string;
  title: string;
  severity: 'EXTREME' | 'HIGH' | 'MODERATE' | 'INFORMATIONAL';
  category: string;
  location: Coordinates;
  affected_radius_km: number;
  message: string;
  issued_at: string;
  expires_at: string;
  source: string;
  is_demo: boolean;
}

export interface AgentTrace {
  agent_name: string;
  status: 'COMPLETED' | 'FALLBACK' | 'SKIPPED';
  execution_time_ms: number;
  data_source: string;
  summary: string;
}

export interface EvidenceDetails {
  intent_detected: string;
  datasets_used: string[];
  timestamps: Record<string, string>;
  deterministic_score: number;
  risk_factors: Record<string, any>;
  observed_vs_forecast: string;
  demo_vs_live: string;
  agent_reasoning_flow: string[];
}

export interface ChatMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  timestamp: string;
  risk_level?: string;
  safety_verdict?: string;
  evidence?: EvidenceDetails;
  traces?: AgentTrace[];
  pfzs?: PFZZone[];
  route?: RouteComparison;
}

export interface ChatResponse {
  direct_answer: string;
  risk_level: 'LOW' | 'MODERATE' | 'HIGH' | 'EXTREME';
  safety_verdict: 'SAFE' | 'SAFE_WITH_CAUTION' | 'UNSAFE' | 'HAZARDOUS';
  recommendation: string;
  conditions_summary?: Record<string, any>;
  evidence: EvidenceDetails;
  agent_traces: AgentTrace[];
  active_map_layers: string[];
  suggested_queries: string[];
  relevant_pfz?: PFZZone[];
  route_comparison?: RouteComparison;
  alerts?: MarineAlert[];
  focus_location: Coordinates;
}

export interface DataSourceInfo {
  id: string;
  name: string;
  organization: string;
  dataset_name: string;
  parameters: string;
  status: 'LIVE' | 'ACTIVE_DEMO' | 'STANDBY';
  is_demo: boolean;
  last_update: string;
  update_frequency: string;
  description: string;
  official_portal: string;
  config_env_var: string;
}

export interface CoastalPreset {
  id: string;
  name: string;
  state: string;
  region: string;
  latitude: number;
  longitude: number;
  harbor: string;
  key_species: string[];
}
